### ISSUE TITLE
Does tagging and publishing update the correlation table? 

### ISSUE THREAD

**Issue by github-germ** [MISP/MISP#5204]
2.4.114

I've noticed that tagging attributes in an event with a large amount of attributes with most of them having correlation enabled that each call to `ExpandedPyMISP.tag` can take upwards of a minute to return. Same for publishing the event once all the tagging is complete.

I also think I am seeing the `correlations` table being updated  as part of the tagging, e.g.

```
UPDATE `misp`.`correlations` AS `Correlation`  SET `Correlation`.`date` = '2019-01-29'  WHERE `Correlation`.`event_id` = 41121
```

I'd be interested in learning how tags are related to correlations. Our `correlations` table is quite large, but wasn't sure why it would be updated when an attribute is tagged. I can understand that the `UPDATE` about could be slow due to table size; however, the `date` hasn't changed (i.e. that is the date before the tagging took place).

Is there something I can to speed up `tag` and `publish` in this use case?

THANKS!!

Btw, I had requested bulk tagging a while ago in MISP/PyMISP#337, that could help this use case if it gets implemented one day.

**Comment by mokaddem** [comment]
Hey,
How are you tagging your attributes (PyMISP, API, GUI), if PyMISP or API, can you provide the function/endpoint?

Anyway, what certainly happens is that if you tag an Attribute, it will update its timestamp and trigger post update hooks (ZMQ, correlation, ...).
So yeah, disabling correlation should speed up the tagging.

**Comment by mokaddem** [comment]
As a feature request, we could whitelist some actions to not re-trigger the correlation.
- Adding tags
- Adding sightings
- switching ``to_ids`` flag
- Adding comment

Or do it the other way around,
1. check if the attribute is a new attribute or not in the aftersave
2. check if type/category/value1/value2 were modified

**Comment by github-germ** [comment]
> Hey,

Thx for the followup!!

> How are you tagging your attributes (PyMISP, API, GUI), if PyMISP or API, can you provide the function/endpoint?

* `ExpandedPyMISP.tag`

> Anyway, what certainly happens is that if you tag an Attribute, it will update its timestamp and trigger post update hooks (ZMQ, correlation, ...).

* Updating the attribute timestamp certainly understandable.
* Why is `corrrelations` updated? OK, I see `correlations.date`; however, if I'm right, tags are not correlated, so from a correlations perspecitive isn't true that nothing changed?

> So yeah, disabling correlation should speed up the tagging.

Let me play with that and report back. Thanks for confirming that as a solution.

**Comment by github-germ** [comment]
OK, so disabling correlations for the event helps, but not quite enough. Seems like that toggle causes the server-side to step thru all attributes in that event in the `correlations` table.

e.g.
```
INSERT INTO `misp`.`correlations` (`value`, `1_event_id`, `1_attribute_id`, `event_id`, `attribute_id`, `org_id`, `distribution`, `a_distribution`, `sharing_group_id`, `a_sharing_group_id`, `date`, `info`) VALUES ('2xo3pb9ak3.com', 98802, 23993122, 137155, 49596585, 33, 0, 0, 0, 0, '2019-09-23', 'All current domains belonging to known malicious DGAs feed')
```

This can take a very long time. I believe it will still be better to wait for that vs. having every tag amnd publish continue to be slow.

I do like to suggestion of a feature request for some actions to not trigger revising the `correlations` table.
* Might that be implementable somewhat quickly?
* Do I need to create another issue making that request?

Thanks!!

**Comment by iglocska** [comment]
> OK, so disabling correlations for the event helps, but not quite enough. Seems like that toggle causes the server-side to step thru all attributes in that event in the `correlations` table.
> 
> e.g.
> 
> ```
> INSERT INTO `misp`.`correlations` (`value`, `1_event_id`, `1_attribute_id`, `event_id`, `attribute_id`, `org_id`, `distribution`, `a_distribution`, `sharing_group_id`, `a_sharing_group_id`, `date`, `info`) VALUES ('2xo3pb9ak3.com', 98802, 23993122, 137155, 49596585, 33, 0, 0, 0, 0, '2019-09-23', 'All current domains belonging to known malicious DGAs feed')
> ```
> 
> This can take a very long time. I believe it will still be better to wait for that vs. having every tag amnd publish continue to be slow.
> 
> I do like to suggestion of a feature request for some actions to not trigger revising the `correlations` table.
> 
> * Might that be implementable somewhat quickly?
> * Do I need to create another issue making that request?
> 
> Thanks!!

It might if you want to implement it quickly. Pull requests are always welcome. Other than that not very quickly no, we have a bunch of higher priority tasks in the pipe.

**Comment by packet-rat** [comment]
Note:  These extraneous processing issues are breaking issues for us.  Correlations across intelligence feeds are key to delivering value.

**Comment by iglocska** [comment]
Did two separate passes on this since yesterday:

1. tagging and a bunch of other tasks will no longer re-correlate attributes. In order for an attribute modification to qualify for recorrelation, it needs to have its value, type or "disable_correlation" field modified.
2. I've looked at the publishing process and identified one stupid mistake that could impede performance. We were accidentally loading a bit too much of the event to simply set the new publish timestamp (loading the attributes, objects, etc) - I've changed this which should make things go faster.

For 2. there is a massive caveat, there are certain functionalities that will be extremely heavy on a constant republishing:
1. If you are synchronising with other instances, even a small modification followed by a republish will load the full event and push it to remote instances.
2. Having the ZMQ/Kafka channels enabled will load the full events and push it down those avenues, be careful with enabling those features if you want to frequently republish.


Las but not least, from what I've heard from @Rafiot - your use-case involves a constant re-tagging of attributes. Did you consider using local tags? These do not unpublish the event and are thus MUCH lighter on the workload.
